Add LetsEncrypt to all web virtualhosts

domcleal opened the pull request:
HTTPS vhosts are all disabled by default via $https parameters, which
allows for bootstrapping when the certificate is not yet available.

On the first run, LE will run using webroot mode and will retrieve a
cert, authenticating using files served by the HTTP vhosts.  On
subsequent runs, $https can be enabled and the HTTPS vhosts will be
configured with the cert.  The letsencrypt module sets up a cronjob for
automatic renewals via the webroot.

A single certificate with SANs is used for all virtual hosts, which even
avoids the need for SNI.

Requires EL7, so post-[migration to web02.theforeman.org](https://groups.google.com/forum/#!topic/foreman-dev/8-HHA6uTGfs).

Once HTTPS is functioning then we can look at migrating all of the various pointers, e.g. documentation, redirecting HTTP requests, yum/apt configs.

GregSutcliffe commented:
+1 Looks sane. I was playing with the LE puppet module myself the other week (on Deb) and it works nicely. My confidence in blind-merging things to -infra has degraded somewhat recently, but I can merge it on Friday ahead of next weeks deadline if there are no objections.

domcleal commented:
Please wait until _after_ the web server migration next week, as doing it before will just cause errors on the EL6 web01.

GregSutcliffe commented:
Ah yes of course. In that case, generic +1, and let's revisit after the cutover.

timogoebel commented:
@domcleal : How are you planning to do the auto renewal of the certificates? Let's Encrypt certificates are only valid for 90 days...

domcleal commented:
@timogoebel letsencrypt::certonly installs a cronjob to perform the renewal.  Thanks for pointing this out though, as I forgot to set `manage_cron => true` - now fixed.  Also fixed a missing require from the HTTPS vhosts on letsencrypt::certonly.

timogoebel commented:
@domcleal : I was unaware the module supports that now. Great. However, with a webroot install you need to restart/reload services (e.g. httpd) after successful renewal. I think, that's not supported by the module right now, is it?

domcleal commented:
Great point, I need to have a think about that.  The graceful weekly restart from logrotate probably isn't going to be enough either, it'll need to be a full restart.

There's development in the LE client on a "renew" command that will indicate whether a cert was renewed or not to make this easier for scripting, but that's due in the next/a future release so not of any immediate use.

GregSutcliffe commented:
I guess the fallback position is to take control of the cron ourselves and add a restart check if it changes anything. Ugly though.

timogoebel commented:
This is how I have done it for my private server, maybe it helps. Basically it just restarts services in the cron-command. It's not rocket science, but it works:
This btw solves another problem: Apache won't start when it can't find the certificates, but LE webroot needs apache started in order to obtain the certificates from ca.

```puppet
# == Class: service_webserver::certificate
#
# Sets up an auto generated certificate
#
define service_webserver::certificate (
  $ssl_domains,
  $docroot,
  $vhostroot,
  $ssl_path,
  $ssl_key,
  $ssl_cert,
  $ssl_chain,
  $auto_renewal = true,
  $reload_services = [],
) {

  # TODO: This can only run after apache has been set up with a vhost on 80
  letsencrypt::certonly { $name:
    domains         => $ssl_domains,
    plugin          => 'webroot',
    additional_args => ["--webroot-path ${docroot}"],
  }

  $primary_ssl_domain = $ssl_domains[0]
  $letsencrypt_ssl_path = "/etc/letsencrypt/live/${primary_ssl_domain}"
  $letsencrypt_ssl_key = "${letsencrypt_ssl_path}/privkey.pem"
  $letsencrypt_ssl_cert = "${letsencrypt_ssl_path}/cert.pem"
  $letsencrypt_ssl_chain = "${letsencrypt_ssl_path}/chain.pem"

  $ssl_csr = "${ssl_path}/csr.pem"

  $stub_ssl_path = "${vhostroot}/certificates/stubs"
  $stub_ssl_key = "${stub_ssl_path}/privkey.pem"
  $stub_ssl_cert = "${stub_ssl_path}/cert.pem"
  $stub_ssl_chain = "${stub_ssl_path}/chain.pem"
  $stub_ssl_csr = "${stub_ssl_path}/csr.pem"

  exec { "make_csr_and_key_${name}":
    command     => "openssl req -batch -new -newkey rsa:4096 -nodes -out ${stub_ssl_csr} -keyout ${stub_ssl_key} -subj '/C=DE/ST=BW/L=Karlsruhe/O=Timo Goebel/OU=Timo Goebel/CN=${primary_ssl_domain}'",
    path        => '/usr/bin',
    environment => 'HOME=/root',
    cwd         => $stub_ssl_path,
    creates     => $stub_ssl_key,
  }

  exec { "sign_csr_${name}":
    command     => "openssl x509 -in ${stub_ssl_csr} -out ${stub_ssl_cert} -req -signkey ${stub_ssl_key} -days 999",
    path        => '/usr/bin',
    environment => 'HOME=/root',
    cwd         => $stub_ssl_path,
    creates     => $stub_ssl_cert,
    require     => Exec["make_csr_and_key_${name}"],
  }

  file { $ssl_key:
    ensure  => link,
    target  => $stub_ssl_key,
    replace => false,
    require => Exec["make_csr_and_key_${name}"],
  }

  file { $ssl_cert:
    ensure  => link,
    target  => $stub_ssl_cert,
    replace => false,
    require => Exec["sign_csr_${name}"],
  }

  file { $ssl_chain:
    ensure  => link,
    target  => $stub_ssl_cert,
    replace => false,
    require => Exec["sign_csr_${name}"],
  }

  File[[$ssl_key, $ssl_cert, $ssl_chain]] -> Apache::Vhost[$name]
  File[[$ssl_key, $ssl_cert, $ssl_chain]] ~> Class['::apache::service']

  exec { "switch_letsencrypt_certs_${name}":
    command => "ln -snf ${letsencrypt_ssl_key} ${ssl_key} && ln -snf ${letsencrypt_ssl_cert} ${ssl_cert} && ln -snf ${letsencrypt_ssl_chain} ${ssl_chain}",
    onlyif  => "test -h ${letsencrypt_ssl_key} && test -h ${letsencrypt_ssl_cert} && test -h ${letsencrypt_ssl_chain}",
    unless  => "test \"$(readlink ${ssl_key})\" -ef \"${letsencrypt_ssl_key}\" && test \"$(readlink ${ssl_cert})\" -ef \"${letsencrypt_ssl_cert}\" && test \"$(readlink ${ssl_chain})\" -ef \"${letsencrypt_ssl_chain}\"",
    require => [File[[$ssl_key, $ssl_cert, $ssl_chain]], Letsencrypt::Certonly[$name]],
    notify  => Class['::apache::service'],
  }

  if $auto_renewal {

    include ::apache::params

    $letsencrypt_path = $::letsencrypt::path
    $plugin = 'webroot'
    $additional_args = ["--webroot-path ${docroot}"]
    $default_reload_services = [$::apache::service_name]
    $real_reload_services = concat($default_reload_services, $reload_services)
    $restart_services = ['dovecot']
    $renew_command = inline_template('<%= @letsencrypt_path %>/letsencrypt-auto --renew certonly -a <%= @plugin %> -d <%= @ssl_domains.join(" -d ")%><% if @additional_args %> <%= @additional_args.join(" ") %><%end%><% if @real_reload_services %><% @real_reload_services.each do |service| %> && systemctl <%= @restart_services.include?(service) ? "restart" : "reload" %> <%= service %><%end%><%end%>')

    cron { "letsencrypt-renewal-${name}":
      command  => $renew_command,
      user     => 'root',
      hour     => 2,
      minute   => fqdn_rand(60),
      monthday => '*/9',
    }
  }
}
```
